File: src/summarizer.ts
import { client, countTokens, MODEL } from "./anthropic.js";
import { config } from "./config.js";
import { chunkText } from "./chunking.js";
import {
  buildMapSystemPrompt,
  buildMapUserPrompt,
  buildReduceSystemPrompt,
  buildReduceUserPrompt,
  buildSystemPrompt,
  buildUserPrompt,
} from "./prompts.js";
import type { SummarizeOptions, StreamEvent } from "./types.js";

type Emit = (event: StreamEvent) => void;

const STREAM_MAX_TOKENS = 64_000; // generous ceiling for streamed output
const MAP_MAX_TOKENS = 8_000; // per-chunk condensations stay compact

/** Pull the concatenated text out of a non-streaming message response. */
function textOf(message: { content: Array<{ type: string }> }): string {
  return (message.content as Array<{ type: string; text?: string }>)
    .filter((b) => b.type === "text")
    .map((b) => b.text ?? "")
    .join("");
}

/**
 * Summarize `opts.text`, streaming the result out through `emit`. Chooses a
 * single-pass request when the input fits comfortably, and a chunked map-reduce
 * pass when it does not.
 */
export async function summarizeStreaming(
  opts: SummarizeOptions,
  emit: Emit,
): Promise<void> {
  const system = buildSystemPrompt(opts);
  const user = buildUserPrompt(opts.text);

  emit({ type: "status", message: "Counting tokens…" });
  const inputTokens = await countTokens(system, user);

  if (inputTokens <= config.chunkThresholdTokens) {
    emit({ type: "meta", inputTokens, strategy: "single-pass" });
    await singlePass(system, user, inputTokens, emit);
  } else {
    await mapReduce(opts, inputTokens, emit);
  }
}

async function singlePass(
  system: string,
  user: string,
  inputTokens: number,
  emit: Emit,
): Promise<void> {
  emit({ type: "status", message: "Generating summary…" });

  const stream = client.messages.stream({
    model: MODEL,
    max_tokens: STREAM_MAX_TOKENS,
    output_config: { effort: "medium" },
    system,
    messages: [{ role: "user", content: user }],
  });

  for await (const event of stream) {
    if (
      event.type === "content_block_delta" &&
      event.delta.type === "text_delta"
    ) {
      emit({ type: "delta", text: event.delta.text });
    }
  }

  const final = await stream.finalMessage();
  emit({
    type: "done",
    outputTokens: final.usage.output_tokens,
    inputTokens,
  });
}

async function mapReduce(
  opts: SummarizeOptions,
  inputTokens: number,
  emit: Emit,
): Promise<void> {
  const chunks = chunkText(opts.text, config.charsPerChunk);
  emit({
    type: "meta",
    inputTokens,
    strategy: "map-reduce",
    chunks: chunks.length,
  });
  emit({
    type: "status",
    message: `Document is large (~${inputTokens.toLocaleString()} tokens). Splitting into ${chunks.length} sections.`,
  });

  // MAP: condense each chunk. Runs at low effort — these are throwaway
  // intermediate summaries, so favor speed and cost.
  const mapSystem = buildMapSystemPrompt();
  const sectionSummaries: string[] = [];
  for (let i = 0; i < chunks.length; i++) {
    emit({
      type: "status",
      message: `Summarizing section ${i + 1} of ${chunks.length}…`,
    });
    const res = await client.messages.create({
      model: MODEL,
      max_tokens: MAP_MAX_TOKENS,
      output_config: { effort: "low" },
      system: mapSystem,
      messages: [
        { role: "user", content: buildMapUserPrompt(chunks[i], i + 1, chunks.length) },
      ],
    });
    sectionSummaries.push(textOf(res));
  }

  // REDUCE: synthesize the final, formatted summary from the section summaries,
  // streaming the result to the client.
  emit({ type: "status", message: "Synthesizing final summary…" });
  const reduceSystem = buildReduceSystemPrompt(opts);
  const reduceUser = buildReduceUserPrompt(sectionSummaries);

  const stream = client.messages.stream({
    model: MODEL,
    max_tokens: STREAM_MAX_TOKENS,
    output_config: { effort: "medium" },
    system: reduceSystem,
    messages: [{ role: "user", content: reduceUser }],
  });

  for await (const event of stream) {
    if (
      event.type === "content_block_delta" &&
      event.delta.type === "text_delta"
    ) {
      emit({ type: "delta", text: event.delta.text });
    }
  }

  const final = await stream.finalMessage();
  emit({
    type: "done",
    outputTokens: final.usage.output_tokens,
    inputTokens,
  });
}

// ---------------------------------------------------------------------------
// Structured extraction: a non-streaming request that returns machine-readable
// JSON. Demonstrates coaxing structured output from the model via a strict
// prompt contract, then parsing it defensively.
// ---------------------------------------------------------------------------

export interface StructuredSummary {
  title: string;
  summary: string;
  keyPoints: string[];
  decisions: string[];
  actionItems: string[];
  entities: string[];
}

const EXTRACT_SYSTEM = [
  "You extract structured information from a document and return it as JSON.",
  "Respond with a SINGLE JSON object and nothing else — no markdown fences, no prose.",
  "The object must match exactly this shape:",
  "{",
  '  "title": string,            // a concise title for the document',
  '  "summary": string,          // 2-4 sentence overview',
  '  "keyPoints": string[],      // most important takeaways, ordered by importance',
  '  "decisions": string[],      // decisions made or proposed (empty array if none)',
  '  "actionItems": string[],    // concrete next steps, each starting with a verb (empty if none)',
  '  "entities": string[]        // notable people, orgs, products, or places mentioned',
  "}",
  "Base every field strictly on the document; never invent content. Use empty arrays where nothing applies.",
].join("\n");

/** Extract a structured summary object from text (non-streaming). */
export async function extractStructured(
  text: string,
): Promise<StructuredSummary> {
  const res = await client.messages.create({
    model: MODEL,
    max_tokens: 8_000,
    output_config: { effort: "medium" },
    system: EXTRACT_SYSTEM,
    messages: [
      {
        role: "user",
        content: [
          "Extract structured information from the document below.",
          "",
          "<document>",
          text,
          "</document>",
        ].join("\n"),
      },
    ],
  });

  return parseStructured(textOf(res));
}

/** Parse the model's JSON defensively, tolerating stray fences or surrounding text. */
function parseStructured(raw: string): StructuredSummary {
  let jsonText = raw.trim();

  // Strip ```json ... ``` fences if the model added them despite instructions.
  const fence = jsonText.match(/```(?:json)?\s*([\s\S]*?)```/);
  if (fence) jsonText = fence[1].trim();

  // Otherwise, grab the outermost {...} span.
  if (!jsonText.startsWith("{")) {
    const start = jsonText.indexOf("{");
    const end = jsonText.lastIndexOf("}");
    if (start !== -1 && end !== -1) jsonText = jsonText.slice(start, end + 1);
  }

  let parsed: Partial<StructuredSummary>;
  try {
    parsed = JSON.parse(jsonText);
  } catch {
    throw new Error("Model did not return valid JSON for structured extraction.");
  }

  const asStringArray = (v: unknown): string[] =>
    Array.isArray(v) ? v.filter((x) => typeof x === "string") : [];

  return {
    title: typeof parsed.title === "string" ? parsed.title : "Untitled",
    summary: typeof parsed.summary === "string" ? parsed.summary : "",
    keyPoints: asStringArray(parsed.keyPoints),
    decisions: asStringArray(parsed.decisions),
    actionItems: asStringArray(parsed.actionItems),
    entities: asStringArray(parsed.entities),
  };
}

File: app/summarizer.py
"""Summarization orchestration: single-pass vs. map-reduce, streaming, structured output."""

from __future__ import annotations

from typing import AsyncIterator

from .chunking import chunk_text
from .config import config
from .llm import MODEL, count_tokens, get_client
from .models import StructuredSummary, SummarizeRequest
from . import prompts

STREAM_MAX_TOKENS = 4_096  # generous ceiling for a streamed summary
MAP_MAX_TOKENS = 1_500  # per-chunk condensations stay compact
SUMMARY_TEMPERATURE = 0.3  # focused, low-variance summaries

# An event is a JSON-serializable dict matching the frontend's StreamEvent union.
Event = dict


async def summarize_streaming(opts: SummarizeRequest) -> AsyncIterator[Event]:
    """Summarize ``opts.text``, yielding stream events. Chooses a single-pass
    request when the input fits comfortably, and a chunked map-reduce pass when
    it does not."""
    system = prompts.build_system_prompt(opts)
    user = prompts.build_user_prompt(opts.text)

    yield {"type": "status", "message": "Counting tokens…"}
    input_tokens = count_tokens(system, user)

    if input_tokens <= config.chunk_threshold_tokens:
        yield {"type": "meta", "inputTokens": input_tokens, "strategy": "single-pass"}
        async for ev in _single_pass(system, user, input_tokens):
            yield ev
    else:
        async for ev in _map_reduce(opts, input_tokens):
            yield ev


async def _stream_summary(
    messages: list[dict], input_tokens: int
) -> AsyncIterator[Event]:
    """Stream one chat completion, yielding text deltas then a 'done' event."""
    stream = await get_client().chat.completions.create(
        model=MODEL,
        messages=messages,
        temperature=SUMMARY_TEMPERATURE,
        max_tokens=STREAM_MAX_TOKENS,
        stream=True,
        stream_options={"include_usage": True},
    )

    output_tokens = 0
    async for chunk in stream:
        if chunk.choices:
            text = chunk.choices[0].delta.content
            if text:
                yield {"type": "delta", "text": text}
        # The final chunk carries usage (choices is empty there).
        if chunk.usage:
            output_tokens = chunk.usage.completion_tokens

    yield {"type": "done", "outputTokens": output_tokens, "inputTokens": input_tokens}


async def _single_pass(
    system: str, user: str, input_tokens: int
) -> AsyncIterator[Event]:
    yield {"type": "status", "message": "Generating summary…"}
    async for ev in _stream_summary(
        [
            {"role": "system", "content": system},
            {"role": "user", "content": user},
        ],
        input_tokens,
    ):
        yield ev


async def _map_reduce(opts: SummarizeRequest, input_tokens: int) -> AsyncIterator[Event]:
    chunks = chunk_text(opts.text, config.chars_per_chunk)
    yield {
        "type": "meta",
        "inputTokens": input_tokens,
        "strategy": "map-reduce",
        "chunks": len(chunks),
    }
    yield {
        "type": "status",
        "message": (
            f"Document is large (~{input_tokens:,} tokens). "
            f"Splitting into {len(chunks)} sections."
        ),
    }

    # MAP: condense each chunk. These are throwaway intermediate summaries, so
    # favor speed with a low temperature and a tight output cap.
    map_system = prompts.build_map_system_prompt()
    section_summaries: list[str] = []
    for i, chunk in enumerate(chunks):
        yield {
            "type": "status",
            "message": f"Summarizing section {i + 1} of {len(chunks)}…",
        }
        res = await get_client().chat.completions.create(
            model=MODEL,
            temperature=0.2,
            max_tokens=MAP_MAX_TOKENS,
            messages=[
                {"role": "system", "content": map_system},
                {
                    "role": "user",
                    "content": prompts.build_map_user_prompt(chunk, i + 1, len(chunks)),
                },
            ],
        )
        section_summaries.append(res.choices[0].message.content or "")

    # REDUCE: synthesize the final, formatted summary from the section summaries,
    # streaming the result to the client.
    yield {"type": "status", "message": "Synthesizing final summary…"}
    async for ev in _stream_summary(
        [
            {"role": "system", "content": prompts.build_reduce_system_prompt(opts)},
            {"role": "user", "content": prompts.build_reduce_user_prompt(section_summaries)},
        ],
        input_tokens,
    ):
        yield ev


# ---------------------------------------------------------------------------
# Structured extraction: a non-streaming request that returns machine-readable
# JSON. Uses OpenAI Structured Outputs (response_format json_schema, strict),
# which guarantees the model's output conforms to the schema.
# ---------------------------------------------------------------------------

_EXTRACT_SYSTEM = (
    "You extract structured information from a document. Base every field strictly "
    "on the document; never invent content. Use empty arrays where nothing applies."
)

# JSON Schema for Structured Outputs. strict mode requires every property to be
# listed in `required` and additionalProperties to be false.
_EXTRACT_SCHEMA = {
    "type": "object",
    "additionalProperties": False,
    "properties": {
        "title": {"type": "string", "description": "A concise title for the document."},
        "summary": {"type": "string", "description": "A 2-4 sentence overview."},
        "keyPoints": {
            "type": "array",
            "items": {"type": "string"},
            "description": "Most important takeaways, ordered by importance.",
        },
        "decisions": {
            "type": "array",
            "items": {"type": "string"},
            "description": "Decisions made or proposed (empty if none).",
        },
        "actionItems": {
            "type": "array",
            "items": {"type": "string"},
            "description": "Concrete next steps, each starting with a verb (empty if none).",
        },
        "entities": {
            "type": "array",
            "items": {"type": "string"},
            "description": "Notable people, orgs, products, or places mentioned.",
        },
    },
    "required": ["title", "summary", "keyPoints", "decisions", "actionItems", "entities"],
}


async def extract_structured(text: str) -> StructuredSummary:
    """Extract a structured summary object from text (non-streaming)."""
    res = await get_client().chat.completions.create(
        model=MODEL,
        temperature=0.2,
        max_tokens=2_000,
        response_format={
            "type": "json_schema",
            "json_schema": {
                "name": "structured_summary",
                "strict": True,
                "schema": _EXTRACT_SCHEMA,
            },
        },
        messages=[
            {"role": "system", "content": _EXTRACT_SYSTEM},
            {
                "role": "user",
                "content": "\n".join(
                    [
                        "Extract structured information from the document below.",
                        "",
                        "<document>",
                        text,
                        "</document>",
                    ]
                ),
            },
        ],
    )

    raw = res.choices[0].message.content or ""
    try:
        return StructuredSummary.model_validate_json(raw)
    except Exception as exc:  # noqa: BLE001
        raise ValueError(
            "Model did not return valid JSON for structured extraction."
        ) from exc
